lib/view/page/create_idea_page.dart

```diff
@@ -0,0 +1,107 @@
+import 'package:flutter/material.dart';
+import 'package:idea_tracker/view/widget/state_management/base_view.dart';
+import 'package:idea_tracker/controller/page/create_idea_controller.dart';
+
+class CreateIdeaPage extends StatelessWidget {
+  final String title;
+  final _formKey = GlobalKey<FormState>();
+  final _focusNode = FocusNode();
+  void _onFieldSubmitted(String value) {
+    _focusNode.requestFocus();
+  }
+
+  CreateIdeaPage({this.title});
+
+  @override
+  Widget build(BuildContext context) {
+    return BaseView<CreateIdeaController>(
+      builder: (context, controller, child) {
+        return Scaffold(
+          appBar: AppBar(
+            centerTitle: true,
+            title: Text(title),
+            leading: IconButton(
+              icon: Icon(Icons.arrow_back),
+              onPressed: () {
+                Navigator.of(context).pop();
+              },
+            ),
+          ),
+          body: Builder(
+            builder: (context) {
+              // Temporary solution to show some feedback
+              if (controller.ideaCreatedSuccesfully) {
+                WidgetsBinding.instance.addPostFrameCallback((_) {
+                  Scaffold.of(context).showSnackBar(SnackBar(
+                    content: Text("Your idea has been created"),
+                  ));
+                });
+              }
```

Comment:
Use a callback function on the controller (registered in the onControllerReady function) instead of checking a boolean value. Comment if you need more instructions here.

Reply:
Do you mean registering a callback in the controller to trigger the call to `showSnackBar()`?

Reply:
Correct, like this:

`final onIdeaCreated = (idea) {
    WidgetsBinding.instance.addPostFrameCallback((_) {
        Scaffold.of(context).showSnackBar(SnackBar(
            content: Text("Your idea has been created"),
        ));
    });
}

return BaseView<CreateIdeaPageController>(
    onControllerReady: (controller) {
        controller.onIdeaCreated = onIdeaCreated;
    }
    builder: (context, controller, child) {
        // ...
    },
);



// on the controller

class CreateIdeaPageController extends ChangeNotifier {
	String _ideaTitle;
	String _ideaDescription;
    Function(Idea) onIdeaCreated;
    
    //...
    
    void createIdea() {
        // TODO: POST idea throught IdeaService
        Future.delayed(Duration(milliseconds: 1000), () {
          final _idea = Idea(
            id: "1",
            title: _ideaTitle,
            description: _ideaDescription,
            createdAt: DateTime.now().millisecondsSinceEpoch,
          );
          if (onIdeaCreated != null) onIdeaCreated(_idea);

          print(
              "Idea created => id: ${_idea.id} | title: ${_idea.title} | description: ${_idea.description} | createdAt: ${_idea.createdAt} | updatedAt: ${_idea.updatedAt} | votes: ${_idea.votes}");
        });
	}
}`

Reply:
I understand. But it can't be done. To call `showSnackBar`, we need `context` and can't be passed to the controller. Even if we manage to pass context to the controller, calling `showSnackBar` from there implies calling it before `Scaffold` in our Widget tree, so we get the error `setState() or markNeedsBuild() called during build`.

Finally, this approach doesn't break the separation of concerns between business logic and UI?

For those reasons, I ended up doing that validation. I couldn't find a better solution, honestly

Reply:
Hi, this doesn't break separation between logic & ui, it enforces it. The button press calls a function in the controller (logic), which then uses a callback to send information back to the view (ui) when the snackbar should appear.

The snackbar is still shown from the view, not the controller, so context is not an issue.

When the app is further along, these pages will reside within a separate Scaffold (which would contain the bottom navigation bar). You are correct that it would cause an error now... I'll make a quick change to the develop branch you can pull into this so an error would not occur. I'll comment here when that is in.

In case you have any other confusion, I have written this example for you:

```
import 'package:flutter/material.dart';
import 'package:get_it/get_it.dart';
import 'package:is_web_enabled/base_view.dart';

GetIt locator = GetIt.instance;

void setupLocator() {
  locator.registerFactory(() => MainPageController());
  locator.registerFactory(() => SubPageController());
}

void main() {
  setupLocator();
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MainPage(),
    );
  }
}

class MainPage extends StatelessWidget {
  const MainPage({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return BaseView<MainPageController>(
      builder: (context, controller, child) {
        return Scaffold(
          body: SubPage(),
          bottomNavigationBar: BottomNavigationBar(
            backgroundColor: Colors.grey,
            items: [
              BottomNavigationBarItem(icon: Icon(Icons.save), label: "Page 1"),
              BottomNavigationBarItem(
                  icon: Icon(Icons.local_dining), label: "Page 2")
            ],
          ),
        );
      },
    );
  }
}

class MainPageController extends ChangeNotifier {}

class SubPage extends StatelessWidget {
  const SubPage({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final onIdeaCreated = (idea) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        Scaffold.of(context).showSnackBar(SnackBar(
          content: Text(idea),
        ));
      });
    };

    return BaseView<SubPageController>(
      onControllerReady: (controller) {
        controller.onIdeaCreated = onIdeaCreated;
      },
      builder: (context, controller, child) {
        return Scaffold(
          appBar: AppBar(
            title: Text("Title"),
          ),
          body: Center(
            child: TextButton(
              onPressed: controller.doAction,
              child: Text("Do Action"),
            ),
          ),
        );
      },
    );
  }
}

class SubPageController extends ChangeNotifier {
  Function(String) onIdeaCreated;

  doAction() async {
    await Future.delayed(Duration(seconds: 1));

    if (onIdeaCreated != null) onIdeaCreated("success!");
  }
}

```

Reply:
In cases where you need the scaffold within your own view, you can use keys to fetch them.

Reply:
The develop branch now has the 'main page' which houses the parent scaffold. View main_page.dart and load your page(s) in the ideasTab Navigator (created in the build method) for testing.

Reply:
Perfect. Pretty clear. I'll implement it tomorrow. Thanks